custom/ucr_ext/expressions.py

```diff
@@ -54,9 +55,51 @@ class GetCaseFormsByDateSpec(JsonObject):
     form_filter = DefaultProperty(required=False)
 
 
+class GetCaseHistorySpec(JsonObject):
+    type = TypeProperty('ext_get_case_history')
+    case_id_expression = DefaultProperty(required=True)
+
+    def configure(self, case_id_expression, case_forms_expression):
+        self._case_id_expression = case_id_expression
+        self._case_forms_expression = case_forms_expression
+
+    def __call__(self, item, context=None):
+        case_id = self._case_id_expression(item, context)
+        if not case_id:
+            return []
+
+        forms = self._case_forms_expression(item, context)
+
+        cache_key = (self.__class__.__name__, case_id)
+        if context.get_cache_value(cache_key) is not None:
+            return context.get_cache_value(cache_key)
+
+        case_history = []
+        for f in forms:
+            case_history.append(_get_case_block(case_id, f))
+
+        context.set_cache_value(cache_key, case_history)
+        return case_history
+
+
+def _get_case_block(case_id, obj):

```

Comment:
i think you might want to leverage [`_extract_case_blocks`](https://github.com/dimagi/commcare-hq/blob/master/corehq/ex-submodules/casexml/apps/case/xform.py#L346) for this, which would hopefully be more maintainable and better tested (might involve some small refactors to get it in the format you want)